src/uuid/uuidbuilder.rs

```diff
@@ -33,174 +32,196 @@ pub struct UUIDBuilder {
+    pub(crate) fn build_from_instance(&self) -> UUID {
+        self.build_internal()
     }
 
     /// Creates a new UUID for a given timestamp.
     ///
+    /// # Note
+    ///
+    /// We use a compare-and-swap (CAS) technique to attempt to repeatedly build a fresh UUID
+    /// until we succeed. Benchmarks roughly twice as fast as using a Mutex.
+    ///
     /// # Arguments
     ///
     /// * `timestamp` - The timestamp (in milliseconds since UNIX EPOCH) to use.
     ///
     /// # Panics
     ///
     /// * if the given timestamp is greater than 2^48 - 1.
-    pub(crate) fn build_with_instant(&self, timestamp: u64) -> UUID {
-        assert!(
-            timestamp & MAX_TIMESTAMP_MASK == 0,
-            "Timestamp of UUID must not exceed 48 bits"
-        );
-
-        let new_msb = {
+    pub(crate) fn build_internal(&self) -> UUID {
+        loop {
             let current_msb = self.msb.load(Ordering::SeqCst);
 
-            if timestamp == (current_msb >> 16) {
-                if (current_msb & MAX_COUNT) < MAX_COUNT {
-                    self.msb.fetch_add(1, Ordering::SeqCst);
+            let cas_top_of_loop_time = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("current system time is set to a point in time before UNIX Epoch");
+            let cas_top_of_loop_millis = u64::try_from(cas_top_of_loop_time.as_millis())
+                .expect("current system time is set to a point in time too far in the future");
+
+            let current_timestamp = current_msb >> 16;
+            let new_msb;
+
+            if cas_top_of_loop_millis == current_timestamp {
+                // If the timestamp hasn't changed, attempt to increment the counter.
+                let current_counter = current_msb & MAX_COUNT;
+                if current_counter < MAX_COUNT {
+                    // Prepare new msb with incremented counter.
+                    new_msb = current_msb + 1;
                 } else {
                     // this should never happen in practice because we
                     // do not expect any uEntity to emit more than
                     // 4095 messages/ms
                     // so we simply keep the current counter at MAX_COUNT
+                    new_msb = current_msb
                 }
             } else {
-                self.msb.store(timestamp << 16, Ordering::SeqCst);
+                // New timestamp, reset counter.
+                new_msb = (cas_top_of_loop_millis << 16) & BITMASK_CLEAR_VERSION
+                    | crate::uuid::VERSION_CUSTOM;
             }
 
-            // set UUID's version to 'custom'
-            self.msb.load(Ordering::SeqCst) & BITMASK_CLEAR_VERSION | crate::uuid::VERSION_CUSTOM
-        };
-
-        UUID::from_u64_pair(new_msb, self.lsb)
-            .expect("should have been able to create UUID for valid timestamp")
+            // Try to update the MSB atomically.
+            match self.msb.compare_exchange(
```

Comment:
Okay, so we use `compare_exchange` instead of `swap` here to ensure no one modifies the `self.msb`, right?
Brilliant idea!